src/compute/src/render/reduce.rs

```diff
@@ -1141,23 +1123,77 @@ where
-    /// obvious trickery here to limit the memory usage per layer by internally
-    /// holding only the elements that were rejected by this stage. However, the
-    /// output collection maintains the `((key, bucket), (passing value)` for this
-    /// stage.
+    /// Build a bucketed stage fragment that wraps [`Self::build_bucketed_negated_output`], and
+    /// adds validation if `validating` is true. It returns the consolidated inputs concatenated
+    /// with the negation of what's produced by the reduction.
     /// `validating` indicates whether we want this stage to perform error detection
     /// for invalid accumulations. Once a stage is clean of such errors, subsequent
     /// stages can skip validation.
+    fn build_bucketed_stage<S>(
+        &self,
+        aggr_funcs: &Vec<AggregateFunc>,
+        input: &Collection<S, (Row, Row), Diff>,
+        validating: bool,
+    ) -> (
+        Collection<S, (Row, Row), Diff>,
+        Option<Collection<S, DataflowError, Diff>>,
+    )
+    where
+        S: Scope<Timestamp = G::Timestamp>,
+    {
+        let (input, negated_output, errs) = if validating {
+            let (input, reduced) = self
+                .build_bucketed_negated_output::<_, _, _, RowValSpine<Result<Row, Row>, _, _>>(
+                    input,
+                    |v| v.into_owned(),
+                    aggr_funcs.clone(),
+                );
+            let (oks, errs) = reduced
+                .as_collection(|k, v| (k.into_owned(), v.clone()))
+                .map_fallible::<CapacityContainerBuilder<_>, CapacityContainerBuilder<_>, _, _, _>(
+                "Checked Invalid Accumulations",
+                |(hash_key, result)| match result {
+                    Err(hash_key) => {
+                        let mut hash_key_iter = hash_key.iter();
+                        let _hash = hash_key_iter.next();
+                        let key = SharedRow::pack(hash_key_iter);
+                        let message = format!(
+                            "Invalid data in source, saw non-positive accumulation \
+                                         for key {key:?} in hierarchical mins-maxes aggregate"
+                        );
+                        Err(EvalError::Internal(message).into())
+                    }
+                    Ok(values) => Ok((hash_key, values)),
+                },
+            );
+            (input, oks, Some(errs))
+        } else {
+            let (input, reduced) = self
+                .build_bucketed_negated_output::<_, _, _, RowRowSpine<_, _>>(
+                    input,
+                    |v| v.into_owned(),
+                    aggr_funcs.clone(),
+                );
+            let oks = reduced.as_collection(|k, v| (k.into_owned(), v.into_owned()));
+            (input, oks, None)
+        };
+
+        let input = input.as_collection(|k, v| (k.into_owned(), v.into_owned()));
+        let oks = negated_output.negate().concat(&input);
```

Comment:
Random thought: we could avoid the `negate` by pre-negating it in `build_bucketed_negated_output`, which already seems like its name means that it might do this. Perhaps just a `// TODO` comment, unless you are feeling excited!

Reply:
I pulled the negate into the reduction; let's see what happens!

Reply:
Works!